Source/JavaScriptCore/assembler/X86Assembler.h

```diff
@@ -6046,46 +6255,20 @@ class X86Assembler {
-            writer.twoBytesVex(isVEX256, simdPrefix, reg, vvvv);
-            writer.putByteUnchecked(opcode);
-            writer.registerModRM(reg, rm);
-        }
-
-        void vexThreeByteOp(bool isVEX256, OneByteOpcodeID simdPrefix, VexImpliedBytes impliedBytes, bool isW1, ThreeByteOpcodeID opcode, uint8_t laneIndex, RegisterID reg, RegisterID rm)
-        {
-            SingleInstructionBufferWriter writer(m_buffer);
-            writer.threeBytesVex(isVEX256, simdPrefix, impliedBytes, isW1, reg, (RegisterID)0, rm, (RegisterID)0);
-            writer.putByteUnchecked(opcode);
-            writer.registerModRM(reg, rm);
-            writer.putByteUnchecked((uint8_t)laneIndex);
-        }
-
-        void vexThreeByteOp(bool isVEX256, OneByteOpcodeID simdPrefix, VexImpliedBytes impliedBytes, bool isW1, ThreeByteOpcodeID opcode, RegisterID reg, RegisterID rm, RegisterID vvvv)
-        {
-            SingleInstructionBufferWriter writer(m_buffer);
-            writer.threeBytesVex(isVEX256, simdPrefix, impliedBytes, isW1, reg, (RegisterID)0, rm, vvvv);
-            writer.putByteUnchecked(opcode);
-            writer.registerModRM(reg, rm);
-        }
-
-        void vexThreeByteOp(bool isVEX256, OneByteOpcodeID simdPrefix, VexImpliedBytes impliedBytes, bool isW1, TwoByteOpcodeID opcode, RegisterID reg, RegisterID rm, RegisterID vvvv)
+        // Typically, xmm1 = dest, xmm2 = a, xmm3 = b
+        // dest = reg, a = vvvv, b = r/m
+        void vexNdsLigWigThreeByteOp(OneByteOpcodeID simdPrefix, VexImpliedBytes impliedBytes, ThreeByteOpcodeID opcode, RegisterID dest, RegisterID a, RegisterID b)
         {
-            SingleInstructionBufferWriter writer(m_buffer);
-            writer.threeBytesVex(isVEX256, simdPrefix, impliedBytes, isW1, reg, (RegisterID)0, rm, vvvv);
-            writer.putByteUnchecked(opcode);
-            writer.registerModRM(reg, rm);
+            vexNdsLigThreeByteOp(simdPrefix, impliedBytes, VexW::W0, opcode, dest, a, b);
         }
 
-        void vexNdsLigWigThreeByteOp(OneByteOpcodeID simdPrefix, VexImpliedBytes impliedBytes, ThreeByteOpcodeID opcode, RegisterID dest, RegisterID a, RegisterID b)
+        // Typically, xmm1 = dest, xmm2 = a, xmm3 = b, but it depends.
+        // dest = reg, a = vvvv, b = r/m
+        void vexNdsLigThreeByteOp(OneByteOpcodeID simdPrefix, VexImpliedBytes impliedBytes, VexW vexW, ThreeByteOpcodeID opcode, RegisterID dest, RegisterID a, RegisterID b)
         {
             SingleInstructionBufferWriter writer(m_buffer);
-            if (impliedBytes != VexImpliedBytes::TwoBytesOp || regRequiresRex(b))
-                writer.threeBytesVexNds(simdPrefix, impliedBytes, dest, a, b);
-            else
-                writer.twoBytesVex(simdPrefix, a, dest);
+            ASSERT(impliedBytes != VexImpliedBytes::TwoBytesOp);
+            writer.threeBytesVex(/* isVEX256 */ false, simdPrefix, impliedBytes, vexW == VexW::W1, dest, (RegisterID)0, b, a);
```

Comment:
Because `threeBytesVex` function is more aligned to the internal encoding of instruction.
Intel Syntax's thing is also a bit different from how actual Vex instructions are encoded. And `threeBytesVex` is very low level instruction for that.
And this is why we are gating the access to threeBytesVex in this patch. Everything will go through the high-level method, and we touch super low-level information only these places.